packages/eslint-plugin/src/rules/prefer-find.ts

```diff
@@ -0,0 +1,283 @@
+    const checker = services.program.getTypeChecker();
+
+    function parseIfArrayFilterExpression(
+      expression: TSESTree.Expression,
+    ):
+      | { isBracketSyntaxForFilter: boolean; filterNode: TSESTree.Node }
+      | undefined {
+      if (expression.type === AST_NODE_TYPES.SequenceExpression) {
+        // Only the last expression in the (a, b, [1, 2, 3].filter(condition))[0] matters
+        const lastExpression = nullThrows(
+          expression.expressions.at(-1),
+          'Expected to have more than zero expressions in a sequence expression',
+        );
+        return parseIfArrayFilterExpression(lastExpression);
+      }
+
+      if (expression.type === AST_NODE_TYPES.ChainExpression) {
+        return parseIfArrayFilterExpression(expression.expression);
+      }
+
+      // Check if it looks like <<stuff>>(...), but not <<stuff>>?.(...)
+      if (
+        expression.type === AST_NODE_TYPES.CallExpression &&
+        !expression.optional
+      ) {
+        const callee = expression.callee;
+        // Check if it looks like <<stuff>>.filter(...) or <<stuff>>['filter'](...),
+        // or the optional chaining variants.
+        if (callee.type === AST_NODE_TYPES.MemberExpression) {
+          const isBracketSyntaxForFilter = callee.computed;
+          if (
+            isBracketSyntaxForFilter
+              ? getStaticValue(callee.property, globalScope)?.value === 'filter'
+              : callee.property.name === 'filter'
+          ) {
+            const filterNode = callee.property;
+
+            const filteredObjectType = getConstrainedTypeAtLocation(
+              services,
+              callee.object,
+            );
+
+            // As long as the object is an array or an optional chain on a
+            // nullable array, this is an Array.prototype.filter expression.
+            if (isArrayish(filteredObjectType)) {
+              return {
+                isBracketSyntaxForFilter,
+                filterNode,
+              };
+            }
+          }
+        }
+      }
+
+      return undefined;
+    }
+
+    /**
+     * Tells whether the type is a possibly nullable array/tuple or union thereof.
+     */
+    function isArrayish(type: Type): boolean {
+      let isAtLeastOneArrayishComponent = false;
+      for (const unionPart of tsutils.unionTypeParts(type)) {
+        if (
+          tsutils.isIntrinsicNullType(unionPart) ||
+          tsutils.isIntrinsicUndefinedType(unionPart)
+        ) {
+          // ignore
+        } else if (
+          checker.isArrayType(unionPart) ||
+          checker.isTupleType(unionPart)
+        ) {
+          isAtLeastOneArrayishComponent = true;
+        } else {
+          return false;
+        }
+      }
+
+      return isAtLeastOneArrayishComponent;
+    }
+
+    function getObjectIfArrayAtExpression(
+      node: TSESTree.CallExpression,
+    ): TSESTree.Expression | undefined {
+      if (node.arguments.length === 1) {
```

Comment:
[Bug] Sorry, just realizing this now on deeper re-review: `filter` can take a second argument, `thisArg`. So this will need to allow that & have test cases.